### PR TITLE
Add jacobian for hydro coupling source term

### PR DESCRIPTION
**Pull request by pajkosmi**
## Proposed changes

Calculate Jacobian for M1 source terms.  Also update M1 system to use Imex timestepper.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by wthrowe** [comment]
Please rebase now that #6076 is merged.

**Comment by pajkosmi** [comment]
> Please rebase now that #6076 is merged.

rebased

File: src/Utilities/Protocols/StaticReturnApplyable.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    static_assert(
+        std::is_same_v<return_tags, tmpl::remove_duplicates<return_tags>>,
+        "return_tags should not contain duplicates.");
+    // Duplicates in argument_tags are OK.
+
+    static_assert(
+        tmpl::wrap<
+            tmpl::append<
+                tmpl::list<decltype(ConformingType::apply)>,
```
Comment (wthrowe) [inline src/Utilities/Protocols/StaticReturnApplyable.hpp]: Test this with a struct where the `apply` function is a template.  I think the current form will not work, but it can be fixed by making a version of `CREATE_IS_CALLABLE` that requires the function to be static.  (By using `T::METHOD_NAME` instead of `std::declval<T>().METHOD_NAME` in the implementation.)

File: src/Utilities/Protocols/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Distributed under the MIT License.
```
Comment (wthrowe) [inline src/Utilities/Protocols/CMakeLists.txt:1]: Include this file from the parent `CMakeLists.txt`.

File: tests/Unit/Utilities/Protocols/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Distributed under the MIT License.
```
Comment (wthrowe) [inline tests/Unit/Utilities/Protocols/CMakeLists.txt:1]: Include this file from the parent `CMakeLists.txt`.

File: src/Evolution/Executables/RadiationTransport/M1Grey/CMakeLists.txt [hunk 4]
```diff
@@ -24,6 +24,7 @@ target_link_libraries(
   GeneralRelativitySolutions
   Hydro
   Informer
+  Imex
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/CMakeLists.txt:26]: Alphabetize.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp [hunk 5]
```diff
@@ -23,8 +23,20 @@ SPECTRE_TEST_CASE("Unit.RadiationTransport.M1Grey.M1HydroCoupling",
       typename RadiationTransport::M1Grey::ComputeM1HydroCoupling<
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp:23]: Add the sector in this commit and call `TestHelpers::imex::test_sector`.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/ComputeM1HydroCouplingJacobian.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
+# Distributed under the MIT License.
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/ComputeM1HydroCouplingJacobian.py:1]: I'm not convinced this python test is worth much.  It's obviously a copy-paste of the c++ source, so any math errors are going to be the same in both.  The IMEX `test_sector` will test this better.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+/// ::protocols::StaticReturnApplyable.  The `return_tags` of this
+/// struct must be a list of `Tags::Jacobian` tags for the dependence
+/// of the sector's `Tags::Sources` on the sector's tensors.  Portions
+/// of the Jacobian that are zero may be omitted from the list.
+///
+/// The `argument_tags` must not include compute items depending on
+/// tensors in the implicit sector.
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct source_jacobian
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: This isn't showing up in the rendered version.  (There is no `struct source_jacobian` in that file.)

File: src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp [hunk 8]
```diff
@@ -49,8 +54,28 @@ void compute_m1_hydro_coupling_impl(
...
+    const Scalar<DataVector>& closure_factor,
+    const Scalar<DataVector>& comoving_energy_density,
+    const tnsr::i<DataVector, 3>& comoving_momentum_density_spatial,
+    const Scalar<DataVector>& comoving_momentum_density_normal,
+    const Scalar<DataVector>& lapse,
+    const tnsr::ii<DataVector, 3>& spatial_metric,
+    const tnsr::II<DataVector, 3>& inverse_spatial_metric);
 }  // namespace detail
 
+// FIXME remove list?
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp]: Make a decision and do it.

File: src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp [hunk 9]
```diff
@@ -78,10 +103,15 @@ struct ComputeM1HydroCoupling;
  * evolutions.
  */
 template <typename... NeutrinoSpecies>
-struct ComputeM1HydroCoupling<tmpl::list<NeutrinoSpecies...>> {
+struct ComputeM1HydroCoupling<tmpl::list<NeutrinoSpecies...>>
+    : tt::ConformsTo<imex::protocols::ImplicitSource>,
+      tt::ConformsTo<protocols::StaticReturnApplyable> {
+  // FIXME remove tags?
+  //  Tags::M1HydroCouplingNormal<NeutrinoSpecies>...,
+  //  Tags::M1HydroCouplingSpatial<Frame::Inertial, NeutrinoSpecies>...>;
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp]: Remove comment.

File: src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp [hunk 10]
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
...
-                             local_time_stepping, system, volume_dim, true>>>,
+                             local_time_stepping, system, volume_dim, true>,
+                             imex::ImplicitDenseOutput<system>>>,
                      evolution::dg::Actions::ApplyLtsBoundaryCorrections<
                          system, volume_dim, false>>,
           tmpl::list<
               evolution::dg::Actions::ApplyBoundaryCorrectionsToTimeDerivative<
                   system, volume_dim, false>,
               Actions::RecordTimeStepperData<system>,
               evolution::Actions::RunEventsAndDenseTriggers<tmpl::list<>>,
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp]: Remove this action.  You added a copy with the correct arguments two lines below.

File: src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp [hunk 11]
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
...
+              imex::Actions::RecordTimeStepperData<system>,
+              evolution::Actions::RunEventsAndDenseTriggers<
+                  tmpl::list<imex::ImplicitDenseOutput<system>>>,
               Actions::UpdateU<system>>>,
       Actions::CleanHistory<system, local_time_stepping>,
+      imex::Actions::DoImplicitStep<system>,  // FIXME at end?
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp]: One line earlier, I think, actually.

File: src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp [hunk 12]
```diff
@@ -192,22 +200,27 @@ struct EvolutionMetavars {
...
       Limiters::Actions::SendData<EvolutionMetavars>,
       Limiters::Actions::Limit<EvolutionMetavars>,
       Actions::MutateApply<typename RadiationTransport::M1Grey::
-                               ComputeM1Closure<neutrino_species>>,
+                               ComputeM1Closure<neutrino_species>>/*,
       Actions::MutateApply<typename RadiationTransport::M1Grey::
-                               ComputeM1HydroCoupling<neutrino_species>>>>;
+                               ComputeM1HydroCoupling<neutrino_species>>*/>>;
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp]: Remove.

File: src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp [hunk 13]
```diff
@@ -223,10 +236,15 @@ struct EvolutionMetavars {
...
       RadiationTransport::M1Grey::Actions::InitializeM1Tags<system>,
+      Initialization::Actions::InitializeItems<imex::Initialize<system>>,
       Actions::MutateApply<typename RadiationTransport::M1Grey::
                                ComputeM1Closure<neutrino_species>>,
       Actions::MutateApply<typename RadiationTransport::M1Grey::
                                ComputeM1HydroCoupling<neutrino_species>>,
+      //   Actions::MutateApply<
+      //       typename
+      //       RadiationTransport::M1Grey::ComputeM1HydroCouplingJacobian<
+      //           neutrinos::ElectronNeutrinos<1>>>,
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp]: Remove.

File: src/Evolution/Systems/RadiationTransport/M1Grey/System.hpp [hunk 14]
```diff
@@ -84,6 +90,58 @@ struct System<tmpl::list<NeutrinoSpecies...>> {
 
   using inverse_spatial_metric_tag =
       gr::Tags::InverseSpatialMetric<DataVector, 3>;
+
+  template <typename Species>
+  struct ImplicitSector : tt::ConformsTo<imex::protocols::ImplicitSector> {
+    using tensors = tmpl::list<Tags::TildeE<Frame::Inertial, Species>,
+                               Tags::TildeS<Frame::Inertial, Species>>;
+    // using helper = ComputeM1Closure<tmpl::list<Species>>;
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/System.hpp]: Remove all the outdated commented stuff in this file.

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp [hunk 15]
```diff
@@ -9,6 +9,8 @@
 #include <utility>  // IWYU pragma: keep  // for move
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: What is the reason for the changes in this commit?  We don't want to do this if we can avoid it.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp (hunk 15 above)
Comment (pajkosmi) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: In `M1HydroCoupling.hpp` the `return_tags` for the `ComputeM1HydroCoupling` struct and `ComputeM1HydroCouplingJacobian` struct need these tags.  Without these changes, the compiler error says the return tags cannot be found in the databox.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp (hunk 5 above)
Comment (pajkosmi) [reply tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp:23]: Would something like `Test_System.cpp` on the below branch suffice?
https://github.com/pajkosmi/spectre/blob/grey_m1/tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp

Comment (wthrowe) [reply tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp:23]: Yes, something like that.

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp (hunk 15 above)
Comment (wthrowe) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: Those are outputs, though.  Why do you need to compute them?

Comment (pajkosmi) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: We will compute them in the EvolveSystem (below link).  My understanding is, even if they're return_tags, they need to be in the Databox?  In `Initialize.hpp`, I see 
```
using m1_variables_tag = typename System::primitive_variables_tag;
```
Do the tags need to be prepended with `::Source` and `::jacobian` in System.hpp and then added to `primitive_variables_tag`?

https://github.com/sxs-collaboration/spectre/blob/887e81c4a148d04634b913185a2ecd368d142152/src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp#L242

File: src/Utilities/Protocols/StaticReturnApplyable.hpp (hunk 1 above)
Comment (pajkosmi) [reply src/Utilities/Protocols/StaticReturnApplyable.hpp]: edit: After having a conversation with @nilsvu, he mentioned this shouldn't work when BaseTags are present, since they do not know their type.  I am going to proceed by keeping the first three `static_asserts` that check for argument_tags, return_tags, and no duplicate return_tags. That should work independently of templated or nontemplated apply functions.  What are your thoughts?

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/ComputeM1HydroCouplingJacobian.py (hunk 6 above)
Comment (pajkosmi) [reply tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/ComputeM1HydroCouplingJacobian.py:1]: I think the fuzzy testing has a small value to check for edge cases producing 0s in denominators.  Nevertheless, I agree with the similarity between the .py and .cpp code, so the test_sector (mentioned below) will be included.

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp (hunk 15 above)
Comment (wthrowe) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: But *why* are you computing them?  What uses them?

Let me make an analogy with runtime code:
```c++
const auto energy = ...;
Scalar<DataVector> energy_source{};
compute_source(make_not_null(&energy_source), energy);
return energy;
```
I am asking the equivalent of "Why is `energy_source` needed?" and getting the answer "It's a required argument of the the `compute_source` function.".  As far as I can tell, you write to these variables, but never read from them, which doesn't seem useful.

File: src/Utilities/Protocols/StaticReturnApplyable.hpp (hunk 1 above)
Comment (wthrowe) [reply src/Utilities/Protocols/StaticReturnApplyable.hpp]: Good point.  That's sounds good.

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp (hunk 15 above)
Comment (pajkosmi) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: Following the SpECTRE discussion, I understand your question now.  The short answer is we do not need these changes, yet.  Formally, this PR does not couple the neutrinos to the fluid.  So nothing, in this PR, needs those terms.  In the future, they will be needed.   

The reason why I included these changes is I want to see if the changes on the original M1 branch were able to compile.  Specifically, in the EvolveM1Grey file, the initialization and step action calls to HydroCoupling were commented out. 

https://github.com/wthrowe/spectre/blob/b73281b3c6c27e029d55097d4fec1f6144bd04a3/src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp#L433

edit (italics is not needed): _After including them in the Evolve system, I found it could not compile, which led me to discover the tags are needed in the databox.  For this PR, I will not include these changes.  For posterity, (myself mostly) the tags for Jacobians of the stiff hydro coupling sources will need to be included._

Comment (wthrowe) [reply src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: They were commented out on the branch because they were left over from the old non-IMEX version.  They were supposed to be deleted when I cleaned up the branch, not uncommented.

You do not need the sources and jacobians in the DataBox in order to couple to fluids.  You can calculate them during the computation of the hydro sources without permanently storing them.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp (hunk 5 above)
Comment (pajkosmi) [reply tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_M1HydroCoupling.cpp:23]: Test_System.cpp has been added to the branch.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp (hunk 7 above)
Comment (pajkosmi) [reply src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: I changed it from `source_jacobian` to `jacobian`.

**Comment by pajkosmi** [comment]
Ping, I freshly rebased and addressed the reviewer comments.

**Comment by wthrowe** [comment]
Don't delete the `CMakeLists.txt` files, or at least the one in `src`.  That's preventing some of the CI checks from running.

**Comment by wthrowe** [comment]
The `CMakeLists.txt` files you had before the fixup were fine, you were just missing the calls to `add_subdirectory` in the parents.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+#include "Evolution/Systems/RadiationTransport/Tags.hpp"
+#include "Helpers/Evolution/Imex/TestSector.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "PointwiseFunctions/Hydro/Tags.hpp"
+#include "Utilities/Literals.hpp"
+#include "Utilities/TMPL.hpp"
+
+SPECTRE_TEST_CASE("Unit.RadiationTransport.M1Grey.System.Imex",
+                  "[Unit][M1Grey]") {
+  // struct DummySpecies;
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp]: Clean up things like this.

File: src/Utilities/CMakeLists.txt [hunk 17]
```diff
@@ -115,6 +115,7 @@ endif()
 
 add_subdirectory(ErrorHandling)
 add_subdirectory(StdHelpers)
+add_subdirectory(Protocols)
```
Comment (wthrowe) [inline src/Utilities/CMakeLists.txt:116]: Alphabetize.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+  Scalar<DataVector> lorentz_factor{1_st};
+  tenex::evaluate(
+      make_not_null(&lorentz_factor),
+      1.0 / sqrt(1.0 - fluid_velocity(ti::I) * fluid_velocity(ti::J) *
+                           spatial_metric(ti::i, ti::j)));
+  // recalculate tilde e (energy density)
+  tenex::evaluate(make_not_null(&tilde_e),
+                  sqrt(tilde_s(ti::i) * tilde_s(ti::j) *
+                       inverse_spatial_metric(ti::I, ti::J)) +
+                      0.01);  // force nearly thin
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp]: Don't just test the thin limit.

File: tests/Unit/Utilities/Protocols/Test_StaticReturnApplyable.cpp [hunk 19]
```diff
@@ -53,4 +69,7 @@ SPECTRE_TEST_CASE("Unit.Utilities.Protocols.StaticReturnApplyable",
       T1{}, T2{}, T3{}, T4{});
   db::mutate_apply<Applyable>(make_not_null(&box));
   CHECK(true);
```
Comment (wthrowe) [inline tests/Unit/Utilities/Protocols/Test_StaticReturnApplyable.cpp]: Only need one `CHECK(true)`.  It's just there to avoid a catch error when there are no checks.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp [hunk 20]
```diff
@@ -29,7 +29,7 @@ namespace imex::protocols {
 /// tensors in the implicit sector.
 ///
 /// \dontinclude Test_SolveImplicitSector.cpp
-/// \skip struct source_jacobian
+/// \skip struct jacobian
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: This example is showing up as
```c++
    struct jacobian {
      using return_tags = tmpl::list<>;
      using argument_tags = tmpl::list<>;
 
      static void apply() {}
    };
```
which doesn't conform to the protocol and isn't a very useful example in any case.

Actually, it looks like nothing actually uses the new protocols, so that should be fixed too.

**Review by wthrowe — CHANGES_REQUESTED** [review]
You can squash the existing fixups.  I'll take another look at the whole thing after.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp (hunk 20 above)
Comment (pajkosmi) [reply src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: This was partly a case-sensitive issue.  `struct Jacobian` (further above) has an implementation that is not a stub.  I also made sure both `Jacobian` and `source` structs now inherit from `ImplicitSource` and `ImplicitSourceJacobian`.

**Comment by pajkosmi** [comment]
squashed

**Comment by wthrowe** [comment]
You don't seem to have squashed.

**Comment by pajkosmi** [comment]
> You don't seem to have squashed.

edit: see below and ignore italics: _When you say squash, do you mean squash all the fixups into 1 fixup commit (like it is now) or squash the 6 current commits (5 initial commits + 1 fixup commit) into 1 commit?_

**Comment by pajkosmi** [comment]
Regarding my previous comment, now I (think) I understand what you meant by squash.  I have squashed the fixup commits into the previous relevant commits and rebased.

File: src/Evolution/Imex/Protocols/ImplicitSector.hpp [hunk 21]
```diff
@@ -93,7 +93,7 @@ namespace imex::protocols {
 /// jacobian:
 ///
 /// \snippet Test_SolveImplicitSector.cpp source
-/// \snippet Test_SolveImplicitSector.cpp jacobian
+/// \snippet Test_SolveImplicitSector.cpp Jacobian
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSector.hpp:99]: Mention/link to the new protocols from the docs in this file, and check in this protocol's `test` that the sub-structs conform to the new protocols.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+/// struct must be a list of `Tags::Jacobian` tags for the dependence
+/// of the sector's `Tags::Sources` on the sector's tensors.  Portions
+/// of the Jacobian that are zero may be omitted from the list.
+///
+/// The `argument_tags` must not include compute items depending on
+/// tensors in the implicit sector.
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct Jacobian
+/// \until };
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: Use the snippet that is used on the `SolveImplicitSector` protocol.

File: src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp [hunk 23]
```diff
@@ -10,10 +10,15 @@
 #include <tuple>
 
 #include "DataStructures/DataBox/DataBox.hpp"
-#include "DataStructures/Tensor/TypeAliases.hpp"
-#include "Evolution/Systems/RadiationTransport/M1Grey/Tags.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"  // IWYU pragma: keep
+#include "Evolution/Imex/Protocols/ImplicitSource.hpp"
+#include "Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp"
+#include "Evolution/Imex/Tags/Jacobian.hpp"
+#include "Evolution/Systems/RadiationTransport/M1Grey/Tags.hpp"  // IWYU pragma: keep
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/M1HydroCoupling.hpp]: Don't add IWYU pragmas.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+      RadiationTransport::M1Grey::Tags::TildeS<Frame::Inertial, DummySpecies>>(
+      explicit_values);
+  get<0>(tilde_s_vals) = DataVector{0.0};
+  get<1>(tilde_s_vals) = DataVector{0.0};
+  get<2>(tilde_s_vals) = DataVector{0.0};
+
+  TestHelpers::imex::test_sector<sector>(stencil_size, tolerance,
+                                         explicit_values, list_of_values);
+
+  // optically thin regime
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp]: Check an intermediate case, not just perfectly thick and thin.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+  get(emissivity) = DataVector{2.0};
+  Scalar<DataVector> absorption_opacity{};
+  get(absorption_opacity) = DataVector{4.0};
+  Scalar<DataVector> scattering_opacity{};
+  get(scattering_opacity) = DataVector{3.0};
+  // fluid velocity
+  tnsr::I<DataVector, 3> fluid_velocity{};
+  get<0>(fluid_velocity) = DataVector{0.};
+  get<1>(fluid_velocity) = DataVector{0.};
+  get<2>(fluid_velocity) = DataVector{0.};
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp]: Don't set this to 0.

File: tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+  // Metric
+  Scalar<DataVector> lapse{};
+  get(lapse) = DataVector{1.0};
+  tnsr::ii<DataVector, 3> spatial_metric{};
+  get<0, 0>(spatial_metric) = DataVector{1.0};
+  get<0, 1>(spatial_metric) = DataVector{0.0};
+  get<0, 2>(spatial_metric) = DataVector{0.0};
+  get<1, 1>(spatial_metric) = DataVector{1.0};
+  get<1, 2>(spatial_metric) = DataVector{0.0};
+  get<2, 2>(spatial_metric) = DataVector{1.0};
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/RadiationTransport/M1Grey/Test_System.cpp]: Set these to something nontrivial.

File: src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp [hunk 27]
```diff
@@ -197,15 +205,16 @@ struct EvolutionMetavars {
               evolution::dg::Actions::ApplyBoundaryCorrectionsToTimeDerivative<
                   system, volume_dim, false>,
               Actions::RecordTimeStepperData<system>,
-              evolution::Actions::RunEventsAndDenseTriggers<tmpl::list<>>,
+              imex::Actions::RecordTimeStepperData<system>,
+              evolution::Actions::RunEventsAndDenseTriggers<
+                  tmpl::list<imex::ImplicitDenseOutput<system>>>,
               Actions::UpdateU<system>>>,
+      imex::Actions::DoImplicitStep<system>,  // FIXME at end?
```
Comment (wthrowe) [inline src/Evolution/Executables/RadiationTransport/M1Grey/EvolveM1Grey.hpp]: Remove comment.

File: tests/InputFiles/RadiationTransport/M1Grey/ConstantM1.yaml [hunk 28]
```diff
@@ -3,6 +3,7 @@
```
Comment (wthrowe) [inline tests/InputFiles/RadiationTransport/M1Grey/ConstantM1.yaml:3]: Presumably the previous commit doesn't work without these changes, so squash these two commits together.

File: src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp [hunk 29]
```diff
@@ -9,6 +9,8 @@
 #include <utility>
```
Comment (wthrowe) [inline src/Evolution/Systems/RadiationTransport/M1Grey/Initialize.hpp]: Drop this commit.

**Review by wthrowe — CHANGES_REQUESTED** [review]
> Regarding my previous comment, now I (think) I understand what you meant by squash. I have squashed the fixup commits into the previous relevant commits and rebased.

Yes, sorry.  You figured it out.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp (hunk 22 above)
Comment (pajkosmi) [reply src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: I do not see a snippet in `SolveImplicitSector.hpp`.  Do you mean change to 
`/// \snippet Test_SolveImplicitSector.cpp Jacobian`
 (like what's used in `ImplicitSector.hpp`), or a snippet somewhere else?

**Comment by pajkosmi** [comment]
Everything has been addressed/rebased, except the `ImplicitSourceJacobian.hpp` point I need clarification on above.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp (hunk 22 above)
Comment (pajkosmi) [reply src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: Update: after talking at the spectre technical call, I went with the above.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp [hunk 30]
```diff
@@ -28,9 +28,8 @@ namespace imex::protocols {
 /// The `argument_tags` must not include compute items depending on
 /// tensors in the implicit sector.
 ///
-/// \dontinclude Test_SolveImplicitSector.cpp
-/// \skip struct Jacobian
-/// \until };
+/// \snippet Test_SolveImplicitSector.cpp Jacobian
+
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: Remove blank line.

File: tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp [hunk 31]
```diff
@@ -183,11 +183,15 @@ struct InitialGuess {
...
-struct Source : tt::ConformsTo<imex::protocols::ImplicitSource> {
+struct Source : tt::ConformsTo<imex::protocols::ImplicitSource>,
+                tt::ConformsTo<::protocols::StaticReturnApplyable> {
   using return_tags = tmpl::list<::Tags::Source<Var2>, ::Tags::Source<Var3>>;
   using argument_tags = tmpl::list<Var1, Var2, Var3, NonTensor,
                                    RecordPreparersForTest, SomeComputeTagBase>;
 
+  static_assert(
+      tt::assert_conforms_to_v<Source, imex::protocols::ImplicitSource>);
+
```
Comment (wthrowe) [inline tests/Unit/Evolution/Imex/Test_SolveImplicitSector.cpp]: Move these (two more below) outside the structs.

File: src/Evolution/Imex/Protocols/ImplicitSource.hpp [hunk 32]
```diff
@@ -0,0 +1,44 @@
...
+/// ::protocols::StaticReturnApplyable.  The `return_tags` of this
+/// struct must be a list of `::Tags::Source<T>` for every tensor `T`
+/// in the implicit sector.
+///
+/// The `argument_tags` must not include compute items depending on
+/// tensors in the implicit sector.
+///
+/// \dontinclude Test_SolveImplicitSector.cpp
+/// \skip struct source
+/// \until };
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSource.hpp]: Switch this example to the preexisting one like you did for the Jacobian.

File: src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp (hunk 22 above)
Comment (wthrowe) [reply src/Evolution/Imex/Protocols/ImplicitSourceJacobian.hpp]: I meant just `ImplicitSector`, sorry.  You figured it out.

File: src/Evolution/Imex/Protocols/ImplicitSector.hpp [hunk 33]
```diff
@@ -93,7 +94,7 @@ namespace imex::protocols {
...
 struct ImplicitSector {
   template <typename ConformingType>
   struct test {
```
Comment (wthrowe) [inline src/Evolution/Imex/Protocols/ImplicitSector.hpp:102]: The checks that the members conform to the new protocols still need to be added in here.

**Review by wthrowe — CHANGES_REQUESTED** [review]
The test of the system looks good now.

**Comment by pajkosmi** [comment]
Comments addressed

**Comment by wthrowe** [comment]
Looks good.  The test timeouts look unrelated to these changes, so don't worry about them.  Go ahead and squash in the remaining fixups.

**Comment by pajkosmi** [comment]
squashed

**Review by wthrowe — APPROVED** [review]
Test failures look unrelated.
